nuttx-patches/priority_restoration_fix.patch

```diff
@@ -0,0 +1,448 @@
+     }
++#endif
+ 
+   return ret;
+ }
+@@ -284,11 +338,11 @@ static int sem_boostholderprio(FAR struct semholder_s *pholder,
+   if (!sched_verifytcb(htcb))
+     {
+       serr("ERROR: TCB 0x%08x is a stale handle, counts lost\n", htcb);
++      DEBUGASSERT(!sched_verifytcb(htcb));
+       sem_freeholder(sem, pholder);
+     }
+ 
+ #if CONFIG_SEM_NNESTPRIO > 0
+-
+   /* If the priority of the thread that is waiting for a count is greater
+    * than the base priority of the thread holding a count, then we may need
+    * to adjust the holder's priority now or later to that priority.
+@@ -322,6 +376,7 @@ static int sem_boostholderprio(FAR struct semholder_s *pholder,
+               else
+                 {
+                   serr("ERROR: CONFIG_SEM_NNESTPRIO exceeded\n");
++                  DEBUGASSERT(htcb->npend_reprio < CONFIG_SEM_NNESTPRIO);
+                 }
+             }
+ 
+@@ -342,8 +397,16 @@ static int sem_boostholderprio(FAR struct semholder_s *pholder,
+            * saved priority and not to the base priority.
+            */
+ 
+-          htcb->pend_reprios[htcb->npend_reprio] = rtcb->sched_priority;
+-          htcb->npend_reprio++;
++          if (htcb->npend_reprio < CONFIG_SEM_NNESTPRIO)
++            {
++              htcb->pend_reprios[htcb->npend_reprio] = rtcb->sched_priority;
++              htcb->npend_reprio++;
++            }
++          else
++            {
++              serr("ERROR: CONFIG_SEM_NNESTPRIO exceeded\n");
++              DEBUGASSERT(htcb->npend_reprio < CONFIG_SEM_NNESTPRIO);
++            }
+         }
+     }
+ 
+@@ -415,10 +478,10 @@ static int sem_dumpholder(FAR struct semholder_s *pholder, FAR sem_t *sem,
+  * Name: sem_restoreholderprio
+  ****************************************************************************/
+ 
+-static int sem_restoreholderprio(FAR struct semholder_s *pholder,
++static int sem_restoreholderprio(FAR struct tcb_s *htcb,
+                                  FAR sem_t *sem, FAR void *arg)
+ {
+-  FAR struct tcb_s *htcb = (FAR struct tcb_s *)pholder->htcb;
++  FAR struct semholder_s *pholder = 0;
+ #if CONFIG_SEM_NNESTPRIO > 0
+   FAR struct tcb_s *stcb = (FAR struct tcb_s *)arg;
+   int rpriority;
+@@ -435,7 +498,12 @@ static int sem_restoreholderprio(FAR struct semholder_s *pholder,
+   if (!sched_verifytcb(htcb))
+     {
+       serr("ERROR: TCB 0x%08x is a stale handle, counts lost\n", htcb);
+-      sem_freeholder(sem, pholder);
++      DEBUGASSERT(!sched_verifytcb(htcb));
++      pholder = sem_findholder(sem, htcb);
++      if (pholder != NULL)
++        {
++          sem_freeholder(sem, pholder);
++        }
+     }
+ 
+   /* Was the priority of the holder thread boosted? If so, then drop its
+@@ -555,6 +623,20 @@ static int sem_restoreholderprio(FAR struct semholder_s *pholder,
+ }
+ 
+ /****************************************************************************
++ * Name: sem_restoreholderprioall
++ *
++ * Description:
++ *   Reprioritize all holders
```

Comment:
This is confusing. It says `Reprioritize all holders`, but in fact it only restores a single holder. I think this is only meant to be used with `sem_foreachholder` (in which case it does restore all)?